uint8 d1
uint8 d2
